Commerce 4

mattstein opened the pull request:
### Description

See [DOC-220](https://linear.app/craftcms/issue/DOC-220/commerce-4-docs).

Commit:
Start branch.

Commit:
Add pages with dynamic content + fix links.

Commit:
Add Donations page.

Commit:
Fix link.

shinybrad committed:
commerce@91be464

Add 3.4.x min required version

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Copy 3.x → 4.x.

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Update requirements page.

Commit:
Fix link.

shinybrad committed:
commerce@ffd1378

Update `3.x` → `4.x` docs references.

shinybrad committed:
commerce@f4155e4

Fix Twig example indentation.

Commit:
More 3.x → 4.x links.

Commit:
Fewer words, correct version.

shinybrad committed:
commerce@f4155e4

Fix Twig example indentation.

Commit:
Update project config references.

Commit:
Upgrade command.

Commit:
Example.

Commit:
`assets/` → `images/`

Might as well match Craft docs.

Commit:
Draft Upgrading content.

Commit:
wip Upgrading sections

Commit:
Title case, apparently.

https://snook.ca/archives/javascript/ajax_vs_ajax

Commit:
Add support for `commerce4:` link prefix.

Commit:
More updates.

Commit:
Warn.

Commit:
`ShippingAddressZone::getStatesNames()`

Commit:
Address subheadings + start Controllers.

Commit:
Removed element methods.

Commit:
Add more changelog items to be massaged.

lukeholder commented on docs/commerce/4.x/upgrading.md:
```diff
@@ -0,0 +1,308 @@
+
+- better address formatting defaults
+- easier address format customization
+- custom address fields can be managed in field layouts—so no more need for `custom1`, `custom2`, etc.
```
Not sure if should go here, or under a description of what the `php craft commerce/upgrade` command does, but the command takes care of prompting you to choose (and create if needed) the custom fields that the old address data should me migrated to. It will do this for the following old address fields:

```attention
title
address3
businessId
phone
alternativePhone
custom1
custom2
custom3
custom4
notes
```

mattstein replied:
[Described on the console commands page](https://github.com/craftcms/docs/pull/347/commits/79056893443afefb5ea26f3c9bc7ab7b9a2b6e58) and linked to it right from the beginning of the upgrade guide.

lukeholder commented on docs/commerce/4.x/upgrading.md:
```diff
@@ -0,0 +1,308 @@
+
+## Zones
+
+## Subscriptions
```
I think the Plans CP URL changes should be left in the changelog? It's not ready a part of upgrading.

mattstein replied:
👍

lukeholder commented on docs/commerce/4.x/upgrading.md:
```diff
@@ -0,0 +1,308 @@
+- Subscription plans are no longer accessible via old Control Panel URLs.
+
+## Discounts
+
```
`discount.userGroupIds` has been removed and migrated into the discount's user condition builder.

mattstein replied:
👍 Mentioned in the table further down, under the “Models” heading.

shinybrad committed:
commerce@dad7e92

Changelog updates

lukeholder commented on docs/commerce/4.x/upgrading.md:
```diff
@@ -0,0 +1,308 @@
+Before you begin, make sure that:
+
+- you’ve reviewed the changes in Commerce 4 [in the changelog](https://github.com/craftcms/commerce/blob/main/CHANGELOG.md#400) and further down this page
+- your site’s running at least **Craft 4.0**
```
One cannot be on Craft 4 first as you would need to update all plugin versions in composer.json to 4.x at the same time?

mattstein replied:
This was silly. Removed.

lukeholder commented on docs/commerce/4.x/upgrading.md:
```diff
@@ -0,0 +1,308 @@
+2. Edit your project’s `composer.json` to require `"craftcms/commerce": "^4.0.0-beta.1"`.
+3. In your terminal, run `composer update`.
+4. Run `php craft migrate/up --track=plugin:commerce`.
+5. Run `php craft commerce/upgrade`.
```
We need to explain that this command is interactive and it will do the following:

1) Make sure you have already run the standard migration.
2) Let the user interactively map any custom countries to valid 2-letter country codes. (All addresses and zones will be updated from this mapping)
3) If any address in the system has data in any of these  fields: 

```
title
address3
businessId
phone
alternativePhone
custom1
custom2
custom3
custom4
notes
```
then they will be prompted to create a custom field (or choose the handle of an existing custom field). The data in the above fields will be migrated to custom fields.

4) All customers in the commerce system will be converted to users if they do not already have a related user existing.
5) All commerce addresses will be created as elements.
6) User addresses will be linked to the migrated address elements.
7) Order addresses will be linked to the migrated address elements
8) Primary billing and shipping IDs on the user are mapped to the address element IDs
9) Tax and shipping zones conditions are created and migrated from the old country and states lists.
10) Store location is connected to it's address element
11) Order history records are now linked directly to user accounts and not to commerce customers
12) Countries that were previously enabled are migrated to store countries,  and an address condition (Market locations) is created.

mattstein replied:
Should be improved in https://github.com/craftcms/docs/commit/79056893443afefb5ea26f3c9bc7ab7b9a2b6e58, where hopefully my rewording is still accurate.

lukeholder commented on docs/commerce/4.x/upgrading.md:
```diff
@@ -0,0 +1,308 @@
+
+States can no longer be enabled or disabled for selection in dropdown lists, but the new Order Address Condition (**Commerce** → **Store Settings** → **Store**) allows you to set rules for allowed addresses.
+
+This example is configured to only allow orders having addresses in the United States:
```
The example in the image is not a country condition but an administrative area condition with no states/provinces chosen.

mattstein replied:
Fixed in newer example.

lukeholder committed:
Changes

- Mention that administrive area can be a free text input if they want.
- remind customer of `gatewayId` if they are going to submit more that one payment form now.
- all of the isEditable isDeletable methods are now can* methods

shinybrad committed:
commerce@cb44659

Fixed permission issue not showing edit/save buttons on order edit

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Clarify relationsihp.

Commit:
`and` → `where`

Commit:
Finish table.

Commit:
Merge pull request #348 from craftcms/feature/luke-additions

Commerce upgrade notes

Commit:
Fix links.

Commit:
More detail for elements.

Commit:
Mention `commerce-manageCustomers` permission removal.

Commit:
Format changed model bits.

Commit:
user → customer

Commit:
Cover store address.

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
Partial Addresses update.

Commit:
Twig + PHP examples.

Commit:
Attempt to clarify purpose of Store Markets.

Commit:
Hopefully reword for clarity.

Commit:
Update PDF config setting description.

shinybrad committed:
commerce@2aaf0a7

Merge branch '4.0' of https://github.com/craftcms/commerce into 4.0

shinybrad committed:
commerce@efa143a

Merge pull request #2714 from craftcms/feature/com-49-refactor-send-email-call-and-method

lukeholder commented on docs/commerce/4.x/upgrading.md:
````diff
@@ -0,0 +1,387 @@
+```
+```php
+// Craft 3
+$countries = \craft\commerce\Plugin::getInstance()
````
See linear https://linear.app/craftcms/issue/DOC-221#comment-d079e8a5 response for clarification

lukeholder commented on docs/commerce/4.x/upgrading.md:
```diff
@@ -0,0 +1,387 @@
+
+![The Order Address Condition condition builder field, configured with `Administrative Area`, `is one of`, `Australia`, and `Western Australia`.](./images/order-address-condition.png)
+
+While Commerce has removed support for managing custom countries and states, the [`commerce/upgrade`](console-commands.md#commerce-upgrade) command migrates previous custom countries and states into fields on individual addresses and adds rules to zone and store market condition builders to match those custom country and state values.
```
*upgrade command prompts you to map the custom countries to real country codes, and copies the state `abbreviation` (if custom state was created) or the `stateName` entered by the customer to the `administrativeArea` field on addresses, and zones.

mattstein replied:
Put this in the wrong place at first, but I [think I’ve got it](https://github.com/craftcms/docs/commit/151cea0ee45332f9ba0c420948c440faf193af7d).

Commit:
Fix event generator versioning.

Commit:
Remove awkward point.

Commit:
Mention field order.

Commit:
Remove sections.

Commit:
Improve table.

Commit:
Revise “Customer → User Transition”.

Commit:
One more sentence.

Commit:
Meaningful “that”.

Commit:
Add tip for getting all countries.

Commit:
Reword + add warning.

Commit:
Mention `makePrimary*`.

Commit:
Describe `commerce/upgrade` command.

Commit:
`businessName` / `businessTaxId` → `orgnization` / `orgnizationTaxId`.

Commit:
Example.

Commit:
Mention order numbers.

Commit:
Mention state `abbreviation` and `stateName`.

Commit:
Move notes to the right place.

Commit:
Merge branch 'develop' into release/commerce/4.0 + minor revisions.

# Conflicts:
#	docs/.vuepress/theme/components/EventBrowser.vue

Commit:
Clarify custom field names, add “Saving an Address”.

Commit:
Try again.

Commit:
Add `config4` anchor prefix.

Commit:
Why not test it.

Commit:
Update instructions and add deployment consideration warning.

Commit:
Add links.

Commit:
This some sketchy code.

Commit:
Fix typo.

Commit:
Shorter.

Commit:
Still shorter.

Commit:
Don‘t allow scrolling-overflow tables to secretly hold open narrow viewports.

Commit:
Use `getAdministrativeAreasListByCountryCode()`.

Commit:
`elements/save` → `users/save-address`

Commit:
Document `|address` Twig filter.

Commit:
Note primary shipping and billing params.

Commit:
Mention user address field layout.

Commit:
Merge branch 'develop' into release/commerce/4.0

Commit:
More detail for address save + delete actions.

Commit:
Let’s be accurate.
